proposals/0000-language-editions.rst

```diff
@@ -0,0 +1,1144 @@
+  language features that do not get their own extension, just a spot
+  in a bundle). It's also conceivable that the bundles will evolve
+  to encompass more expressive power (such as controlling optimization
+  flags or the meaning of ``import Prelude``).
+
+  Unlike language editions, a user may specify any number of bundles
+  when compiling a file. In order to simplify their processing, a
+  bundle cannot control the interpretation of other command-line
+  flags. Otherwise, though, a bundle can potentially control arbitrary
+  behavior of GHC, just like a language edition. The big difference
+  between them is that a file is compiled against exactly one language
+  edition, but it can be compiled against an arbitrary number of bundles.
+
+- When printing out the namne of a warning flag as part of a warning,
+  we also include any bundle that also controls the warning.
+
+Detailed specification
+~~~~~~~~~~~~~~~~~~~~~~
+
+The text above defines language editions generally. This section
+instantiates the general design with some specifics. These language
+editions affect language features (as described by extension name or
+warning flag in the chart below), as well as error messages (not described
+in any detail). They do not affect other aspects of GHC.
+
+The chart below classifies all current extensions and warnings.
+
+* The GHC2021 column has an X for extensions enabled as part of GHC2021.
+  For warnings, it lists if the warning is part of a current warning bucket.
+
+* The next four columns describe ``Student2024``, ``Stable2024``, ``Experimental2024``, and ``Latest``,
+  respectively. Here is the key for these columns:
+
+  * Y: an extension is turned on
+
+  * blank: the feature is available, but not on
+
+  * N: trying to enable the feature is an error
+
+  * W: the warning warns by default
+
+  * E: the warning is an error by default
+
+  * W!: the warning warns by default and cannot be turned off
+
+  * E!: the warning is an error by default and cannot be turned off or made into a warning
+
+* The next several columns describe semantic bundles of options.
+
+  * ``FancyTypes``: The user should enable this if they want fancy types. By enabling
+    this semantic bundle, the user takes responsibility for understanding type inference
+    and interactions at a deeper level. In an enterprise setting, the user takes the
+    responsibility for training future collaborators in these techniques.
+
+    Includes the following: ``DataKinds``, ``ExistentialQuantification``, ``GADTs``,
+    ``GADTSyntax``, ``ImpredicativeTypes`` (allowing polytypes in visible type applications
+    and in type parameters, no Quick Look), ``QuantifiedConstraints``, ``RankNTypes``,
+    ``RequiredTypeArguments``, ``RoleAnnotations``, ``TypeAbstractions``, ``TypeData``,
+    ``TypeFamilies``, ``-Wterm-variable-capture``.
+
+  * ``DoSyntax``: This enables extra syntactic support around ``do``-notation. Someone
```

Comment:
What is the stability story for bundles? Presumably if we add a new kind of extension that affects do syntax, then we would want to add it to `DoSyntax`, but we don't want to change the meaning of old code. So I think the language edition must also control the meaning of bundles?

Reply:
Of course! But I agree that this is not spelled out in the proposal. The proposed bundles all have the same meaning regardless of the proposed edition. This might not be true in the future. I will clarify.